webhooks: Add a new preset url options, `mapping`.

PieterCK opened the pull request:
This PR adds a "mapping" url option that adds "Matching Zulip channel" option to the stream dropdown widget. When that option is chosen from the dropdown, it adds another parameter to the integration URL -- "&mapping=channels".

This PR temporarily adds a url option to the Airbrake and GitHub integration to test the new feature. They have nothing to do with this option, and the last commit is not meant to be merged.

Technical discussion: [#integrations > Slack's channels_map_to_topics setting. @ 💬](https://chat.zulip.org/#narrow/channel/127-integrations/topic/Slack's.20channels_map_to_topics.20setting.2E/near/2062232)
Design discussion: [#design > UI for integration's "interfaced settings" #33370 @ 💬](https://chat.zulip.org/#narrow/channel/101-design/topic/UI.20for.20integration's.20.22interfaced.20settings.22.20.2333370/near/2123070)

Picks up from #34113.

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
(tested again July 10 2025)
| Dark | Light |
|--------|--------|
| ![image](https://github.com/user-attachments/assets/a3dbcd15-dc38-4d3f-a818-eb7b11dc6e89) | ![image](https://github.com/user-attachments/assets/97790e85-949a-42fe-b717-972d24b8fd70) | 

| When selected |
|--------|
| ![image](https://github.com/user-attachments/assets/9b0e33d8-1933-4b4d-943d-abd4d5fa29e9) | 

| Preset url options doc (updated July 10 2025) |
|--------|
|![image](https://github.com/user-attachments/assets/4451fe8f-fc54-41c9-88a7-d21a4e8c2ca1)| 

### Follow up:
- Update Slack to use the mapping=channels readable parameter, replacing the channels_map_to_topics parameter.
- Update the rest of the bridge-like integrations to use this url option.

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

PieterCK commented on web/src/integration_url_modal.ts:
```diff
@@ -49,6 +62,11 @@ export function show_generate_integration_url_modal(api_key: string): void {
     });
+    const map_channels_option: Option = {
+        name: $t_html({defaultMessage: "Matching Zulip channel"}),
+        unique_id: -2,
```
I didn't end up refactoring the `unique_id` to work nicely with cases like this, where we can have both string and number as the unique ID type. The DropdownWidget module is used widely, so I'm not sure if the risky refactor over correctness for this case is worth blocking this PR.

PieterCK commented:
Fixed a couple of typos and updated the equals icon SVG (still tentative).

timabbott commented on api_docs/incoming-webhooks-walkthrough.md:
```diff
@@ -655,6 +659,45 @@ with a string describing the unsupported event type, like so:
+
+### Preset configuration options
+
+Preset config options unique `WebhookConfigOption` values that come
```
This is missing a word, "are unique"?

timabbott commented on api_docs/incoming-webhooks-walkthrough.md:
````diff
@@ -655,6 +659,45 @@ with a string describing the unsupported event type, like so:
+    ),
+```
+
+Here are the preset config options and their behavior when the integration
````
`integration(s)`

timabbott commented:
I did a quick read and this looks to be well structured. I'll let @Niloth-p do a read and test, but I am optimistic we can integrate this soon.

PieterCK commented:
Updated the SVG to use Lucide's equal icon and fixed some wording in the preset config documentation.

alya commented:
@Niloth-p Please take a look at this one when you get a chance.

PieterCK commented on api_docs/incoming-webhooks-walkthrough.md:
```diff
@@ -655,6 +659,45 @@ with a string describing the unsupported event type, like so:
+# zerver/lib/integrations.py
+from zerver.lib.webhooks.common import PresetConfigOption, WebhookConfigOption
+  # ...
+    WebhookIntegration(
```
Right! changed it to:

```python
# zerver/lib/integrations.py
from zerver.lib.webhooks.common import PresetConfigOption, WebhookConfigOption
  # -- snip --
    WebhookIntegration(
        "github",
        # -- snip --
        config_options=[
            WebhookConfigOption.preset_config(PresetConfigOption.BRANCHES),
        ],
    ),
```

PieterCK commented on api_docs/incoming-webhooks-walkthrough.md:
```diff
@@ -655,6 +659,45 @@ with a string describing the unsupported event type, like so:
+    it will show a list of all the integrations' supported events. The user
+    can then filter which events can trigger a notification.
+
+    Some integrations that uses this are: GitHub, Gitea, GOGS.
```
> ...list all the supported integrations in a single place.

hmm, I'm not sure why that's something we'd want to do. I think the point of giving examples is to give the reader more context on how the setting could be used, so examples of a few main use cases would probably suffice. 

I ended up adding them anyways, just because there's only a handful. Though I think later on it can be a context variable, with a name like `{{branches_option_integrations}}` so that we don't have to manually keep track of them.

PieterCK commented on zerver/lib/webhooks/common.py:
```diff
@@ -54,12 +56,46 @@
+        cls, config: PresetConfigOption, description: str = ""
+    ) -> "WebhookConfigOption":
+        """
+        This WebhookConfigOptions comes with a custom UI and logic in the
```
I think for cases like this, we still include a bit of context in the function, e.g `zerver/actions/message_send.py/create_user_messages`. I updated the function comment to read slightly better:

```python
        """
        This creates `WebhookConfigOptions` for the given `PresetConfigOptions`,
        those are config options that comes with a custom UI and logic in the
        "Generate integration URL" modal.

        See /api/incoming-webhooks-walkthrough#preset-configuration-options
        for more details on this system and what each `PresetConfigOptions` do.
        """
```

PieterCK commented on api_docs/incoming-webhooks-walkthrough.md:
````diff
@@ -697,6 +697,22 @@ that uses them is selected:
+    field and adds `&z_mapping=channels` variable to the integration URL, like so:
+
+    ```
+    {{zulip_url}}/api/v1/external/github?api_key=kT2KTUBJVv0kAyrwY3povICtTaczHQwl&z_mapping=channels
````
Updated it to Slack

PieterCK commented on api_docs/incoming-webhooks-walkthrough.md:
```diff
@@ -655,6 +659,45 @@ with a string describing the unsupported event type, like so:
+that uses them is selected:
+
+  1. **BRANCHES**: Adds a new field to the "Generate integration URL" modal
+    called "Filter events that will trigger notifications?". If the selected,
```
uhoh, right! 😄

PieterCK commented on api_docs/incoming-webhooks-walkthrough.md:
```diff
@@ -655,6 +659,45 @@ with a string describing the unsupported event type, like so:
+with pre-built user interface elements -- new field, validation, etc --
+in the "Generate integration URL" modal. These are useful if the integration
+requires a more elaborate UI in the frontend then what plain `config_options`
+can offer, or if the setting could be used by many integrations.
```
Yeah, that bit reads a bit awkward on second thought. I think I was trying to highlight how if a vanilla `config_option` is generic enough to be used by many integrations, it could instead be turned into a preset config option, so that the same `description` string doesn't have to be written over and over and even be standardised. I removed it for now, since there's currently no such use case.

PieterCK commented on api_docs/incoming-webhooks-walkthrough.md:
```diff
@@ -655,6 +659,45 @@ with a string describing the unsupported event type, like so:
+### Preset configuration options
+
+Preset config options are unique `WebhookConfigOption` values that come
+with pre-built user interface elements -- new field, validation, etc --
```
Good point, I've simplified the description to this:

```
Preset config options are unique `WebhookConfigOption` objects that may come
with more elaborate user interface elements and logic than what a plain
`WebhookConfigOption` can offer.
```

PieterCK commented on web/src/dropdown_widget.ts:
```diff
@@ -31,6 +31,7 @@ export type Option = {
     description?: string;
     is_direct_message?: boolean;
     is_setting_disabled?: boolean;
+    is_preset_config_option?: boolean;
```
Good catch, it definitely is too generic. I ended up not going down the route of adding more fields to the Dropdown module's `Option` object. This is another symptom of using two types (string & number) for `unique_id_type` when it only wants one. So, I decided to look for the unique ID of `map_channels_option`, which is -2. This way, once we do the refactor for `unique_id_type` (if at all), this will also get fixed.

PieterCK commented:
@Niloth-p Thanks for looking through this! I've replied to your comments on how the new changes address them. Other than that, I decided not to make the change encoding the "branches" config option to "z_branches"; I'd think the benefit of having that encoded isn't worth the complexity of adding backwards compatibility logic for all existing integrations. Let me know if there's anything I've missed.

PieterCK commented:
@Niloth-p - I pushed some more updates for this, it's ready for you to take a look again. And I kindly request a `documentation review` from @laurynmm for this!

Changes:
- I moved the new **Configuration option presets** to just below the **Webhooks requiring custom configuration** section, they are in the subsection of Step 3. I didn't end up combining the two sections as per suggested because the latter is still outdated, and I think it's useful to have a link directly to the **Configuration option presets** section. But yeah it doesn't need to be at the end of the page.
- I rewrote the definition for configuration option presets, but yeah it's still rough around the edges. :sweat_smile: 

@zulipbot add "documentation review"

PieterCK commented on api_docs/incoming-webhooks-walkthrough.md:
````diff
@@ -231,6 +231,71 @@ configure:
+    {{zulip_url}}/api/v1/external/github?api_key=kT2KTUBJVv0kAyrwY3povICtTaczHQwl&branches=main%2Cdev
+    ```
+
+    Integration(s) using this: GitHub, Gitea, Gitlab, rhodecode, azuredevops
````
Dropped the (s) and capitalized the first letter of each items:
```
    Integrations using this: GitHub, Gitea, Gitlab, Rhodecode, Azuredevops
    Bitbucket3, Bitbucket2, Gogs.
```

PieterCK commented:
Thanks for the feedback! @Niloth-p. Sorry, I had the wrong version up, should be up to date now (description & commits).

>Please update the screenshot. It seems to have been updated since last review, but does not match the code? I'm not sure which is the latest version because there are certain sections where the screenshot reads better than the text in the code, and certain sections are the other way around.

My bad! I missed pushing the last batch of changes, it seems.

>We could just make it a level 4 heading, and use it as a link?
We're also just using a file link in a docstring, so the section link doesn't work anyways? Or is the link used elsewhere?

I meant for sending the link to this doc in general (like in CZO or to other contributors). I've updated to use level 4 heading for this! We can't link the section anymore, but I guess this is a very easy thing to change if we decide otherwise

> The ignore_private_repositories option - what's the plan for that? Will we be tracking it as a follow-up? Would we be using that as an example for that type of preset config options?

Good idea, added a commit for this: 
`integrations: Refactor integrations to use preset_config.`

>Do we really prefer file links over hyperlinks, when it's not in RTD? I've not heard of this before, so I'd like to check, because I personally prefer hyperlinks.

I think Lauryn can also confirm this. For context, we're talking about the doc string for `WebhookConfigOption.preset_config`:
```python
    @classmethod
    def preset_config(
        cls, config: PresetConfigOption, description: str = ""
    ) -> "WebhookConfigOption":
        """
        This creates pre-configured `WebhookConfigOption` objects. Some presets
        build normal `WebhookConfigOption` objects, while others build special
        `WebhookConfigOption` objects with additional logic and UI elements.

        See /api_docs/incoming-webhooks-walkthrough#configuration-option-presets
        for more details on this system and what each `PresetConfigOptions` do.
        """
```
I'm proposing we use filepath(or relative link) so that it points to the correct version of the doc for the feature.

PieterCK commented on api_docs/incoming-webhooks-walkthrough.md:
```diff
@@ -231,6 +231,75 @@ configure:
+    Integrations using this: GitHub, Gitea, Gitlab, Rhodecode, Azuredevops
+    Bitbucket3, Bitbucket2, Gogs.
+
+  2. **IGNORE_PRIVATE_REPOSITORIES**: Preset for a normal `WebhookConfigData` object
```
Looks better!

PieterCK commented on api_docs/incoming-webhooks-walkthrough.md:
```diff
@@ -231,6 +231,75 @@ configure:
+
+    Integration using this: GitHub
+
+  2. **MAPPING**: Adds a new option to the "Where to send notification?" drop
```
oops! Changed it

PieterCK commented:
Manually tested this again, this is ready for you to take alook again! @Niloth-p

PieterCK commented:
>I assume we use the switch-case instead inside of if(is_preset_config_option(option.key)) , instead of using if(option.key === PresetConfigOption.BRANCHES) and if(option.key === PresetConfigOption.MAPPING), in order to generalize the flow. But, if so, duplicating the code for case PresetConfigOption.IGNORE_PRIVATE_REPOSITORIES feels awkward to me, defeating the purpose of the switch-case that was added.

hmm, yeah I went with that approach because we'd have to include `IGNORE_PRIVATE_REPOSITORIES` in `integration_url_modal.PresetConfigOption`. The duplicate code inside the switch case block is to make it clear that although  `IGNORE_PRIVATE_REPOSITORIES` is a `PresetConfigOption`,  we treat it normally like a plain `ConfigOption` object. I think it'd be abit awkward either way -- from having to leave out `IGNORE_PRIVATE_REPOSITORIES` of `integration_url_modal.PresetConfigOption` or from having a duplicate logic inside the switch case. But by having this duplicate logic for `IGNORE_PRIVATE_REPOSITORIES` we can be clear about what it's suppose to do and don't have to do special casing like:

````javascript
if is_preset_config_option(option.key) && option.key != IGNORE_PRIVATE_REPOSITORIES {
  ...
````
(or something like that)
Would love to hear your suggestion

PieterCK commented:
Right, I reverted the logic in `render_config` to this:

```typescript
if (option.key === PresetConfigOption.BRANCHES) {
    const filter_branches_html =
        render_generate_integration_url_filter_branches_modal();
    $config_element = $(filter_branches_html);
    $config_element.find("#integration-url-all-branches").on("change", () => {
        show_branch_filtering_ui();
    });
} else if (option.key === PresetConfigOption.MAPPING) {
    continue;
} else if (option.validator === "check_bool") {
    const config_html = render_generate_integration_url_config_checkbox_modal({
        key: option.key,
        label: option.label,
    });
    $config_element = $(config_html);
    $config_element
        .find(`#integration-url-${option.key}-checkbox`)
        .on("change", () => {
            update_url();
        });
} else if (option.validator === "check_string") {
    const config_html = render_generate_integration_url_config_text_modal({
        key: option.key,
        label: option.label,
    });
    $config_element = $(config_html);
    $config_element.find(`#integration-url-${option.key}-text`).on("input", () => {
        update_url();
    });
} else {
    continue;
}
```

Thanks, this https://github.com/zulip/zulip/pull/34353#issuecomment-2885902405 was clear. I've also done another manual testing.

PieterCK commented on web/src/integration_url_modal.ts:
```diff
@@ -33,6 +33,13 @@ const config_option_schema = z.object({
 const config_options_schema = z.array(config_option_schema);
 
+const PresetConfigOption = {
+    // Keep this in sync with `PresetConfigOption`
```
Cleared  :+1:

laurynmm commented on zerver/lib/webhooks/common.py:
```diff
@@ -54,12 +56,43 @@
+        """
+        This creates pre-configured `WebhookConfigOption` objects. Some presets
+        build normal `WebhookConfigOption` objects, while others build special
+        `WebhookConfigOption` objects with additional logic and UI elements.
```
The  "normal objects" vs "special objects with additional logic and UI elements" are related to the web-app modal for building the webhook URL, correct? I don't think that's clear in this documentation.

I'd recommend simplifying this to "This creates a pre-configured `WebhookConfigOption` object that's used in various incoming webhook integrations." or something like that.

PieterCK replied:
Okay! That reads better

laurynmm commented on zerver/lib/webhooks/common.py:
```diff
@@ -54,12 +56,43 @@
+        `WebhookConfigOption` objects with additional logic and UI elements.
+
+        See https://zulip.com/api/incoming-webhooks-walkthrough#configuration-option-presets
+        for more details on this system and what each `PresetConfigOptions` do.
```
-> "and what each option does."

laurynmm commented on zerver/lib/webhooks/common.py:
```diff
@@ -54,12 +56,43 @@
+
+        raise AssertionError(
+            _(
+                "Please configure how to build the '{}' option in WebhookConfigOption.preset_config.",
```
How about "Unknown PresetConfigOption: {config}"?

PieterCK replied:
Updated to:
```python
        raise AssertionError(_("Unknown PresetConfigOption: %s") % config)
```

laurynmm commented on zerver/lib/integrations.py:
```diff
@@ -400,9 +400,7 @@ def __init__(self, name: str, *args: Any, **kwargs: Any) -> None:
-        config_options=[
-            WebhookConfigOption(name="branches", description="", validator=check_string)
-        ],
+        config_options=[WebhookConfigOption.preset_config(PresetConfigOption.BRANCHES)],
```
Why are the validators for these changing to `check_bool`? According to the current documentation on `WebhookConfigOption` in the incoming webhook tutorial, that's used to validate the value for the parameter: https://zulip.com/api/incoming-webhooks-walkthrough#webhooks-requiring-custom-configuration. Is that not accurate?

PieterCK replied:
Yeah, that should still be `check_string`. But for cases like `BRANCHES` and `MAPPING` they have custom validation that overrides validation for their  `validator` key.

laurynmm commented on zerver/lib/webhooks/common.py:
```diff
@@ -85,6 +86,12 @@ def preset_config(
+            case PresetConfigOption.IGNORE_PRIVATE_REPOSITORIES:
+                return cls(
+                    name=config.value,
+                    description="Exclude notifications from private repositories",
```
So, if someone passed a description to the classmethod here, it would be ignored, correct? Again, I wonder if we want/need the option to pass a description string to this method.

PieterCK replied:
Correct! https://github.com/zulip/zulip/pull/34353#discussion_r2097061722

laurynmm commented on web/src/integration_url_modal.ts:
```diff
@@ -262,7 +273,15 @@ export function show_generate_integration_url_modal(api_key: string): void {
                         $input_element = $(`#integration-url-${option.key}-text`);
                         const value = $input_element.val();
-                        if (value) {
+                        // If the config option is "branches", ensure the checkbox is unchecked.
```
Didn't think get changed to `check_bool` in a previous commit?

laurynmm commented on zerver/lib/webhooks/common.py:
```diff
@@ -57,8 +57,14 @@
     IGNORE_PRIVATE_REPOSITORIES = "ignore_private_repositories"
+    # Encode the config's name with the "z_" prefix as to not
+    # accidentally trigger a config's custom behavior when
+    # manually declaring new config with identical key name.
```
I don't quite understand this comment. The concern is that an existing webhook may have "mapping" as a parameter?

PieterCK replied:
The concern is that integrations using `WebhookConfigOption` with identical `name` key ("mapping") would unintentionally trigger `PresetConfigOption.MAPPING`s' logic in the modal. Encoding these configs with a "z_" prefix means we don't have to double-check and worry about `name` keys colliding when declaring new `WebhookConfigOption`. The repercussion of said error is fairly negligible, as it should be immediately obvious, so not having this rule is also an option.

laurynmm reviewed (commented):
@PieterCK - Thanks for working on this update! I have a bunch of thoughts on the incoming webhook tutorial documentation updates, but I thought it would be good to hold off on that until some of my thoughts/questions about the implementation changes are cleared up.

Let me know if you have any questions about my comments below!

PieterCK commented on zerver/lib/webhooks/common.py:
```diff
@@ -54,12 +56,43 @@
 
+    @classmethod
+    def preset_config(
+        cls, config: PresetConfigOption, description: str = ""
```
right, not right now but I imagine a preset that allows for some degree of customization of each object. Say, the `MAPPING` option right now adds a new dropdown option with a fixed description, I imagine it's possible to have a similar option where the name/description of the dropdown option is customizable. Such setting would still have to be a `PresetConfigOption` since we need its key to be of unique value.

I removed the `description` parameter for now, as none of the current presets use that mechanism.

PieterCK commented on zerver/lib/integrations.py:
```diff
@@ -458,6 +462,7 @@ def __init__(self, name: str, *args: Any, **kwargs: Any) -> None:
         config_options=[
             WebhookConfigOption.preset_config(PresetConfigOption.BRANCHES),
             WebhookConfigOption.preset_config(PresetConfigOption.IGNORE_PRIVATE_REPOSITORIES),
+            WebhookConfigOption.preset_config(PresetConfigOption.MAPPING),
```
>to test seeing it when other configuration presets were also configured?

Yes!
I've updated the temporary commit message

PieterCK commented on zerver/lib/webhooks/common.py:
```diff
@@ -54,12 +56,43 @@
 OptionalUserSpecifiedTopicStr: TypeAlias = Annotated[str | None, ApiParamConfig("topic")]
 
 
+class PresetConfigOption(str, Enum):
```
Updated to use `StrEnum`

laurynmm replied:
Ah, okay.

PieterCK commented on zerver/lib/webhooks/common.py:
```diff
@@ -54,12 +56,43 @@
     validator: Callable[[str, str], str | bool | None]
 
+    @classmethod
+    def preset_config(
```
Opted for `build_preset_config`

PieterCK commented:
@laurynmm - This is ready for review again. Most of the changes are self-explanatory -- changing strings or variable names as per suggested, I did some refactor here https://github.com/zulip/zulip/pull/34353#discussion_r2100108649 to address the comment in that thread. I've also updated the screenshots in the PR description and did another manual testing: 
![image](https://github.com/user-attachments/assets/938e8a2d-44a4-4ca9-9cd1-9e98befc5f54)
Let me know if I've missed anything!

laurynmm commented:
@PieterCK - Thanks for making those updates and apologies for the delay in getting this reviewed again. I've pushed some further revisions to the documentation changes here (see screenshot below), along with a few other updates detailed below.

**Updates**:
- Squashed your first two commits so that adding the "branches" preset URL configuration all happens in the same commit.
- Squashed the commit with the mapping icon with the commit that added that special preset URL configuration.
- Moved where the `map_channels_option` constant was defined in the integration url modal logic for readability.

@timabbott - Note that the last commit in the series should be dropped before merging. I'm leaving it here in case you'd like to quickly do any Q&A testing when you do your review.

---

**Updated documentation screenshot**:

<img width="771" height="941" alt="Screenshot 2025-07-26 at 22 21 44" src="https://github.com/user-attachments/assets/cf2c3e36-ee07-4b6b-9abe-4705573d85f7" />

timabbott commented:
I dropped the DO NOT MERGE testing commit and merged, thanks for all the work on this @PieterCK, @Niloth-p and @laurynmm!

PieterCK commented:
Okay, I'll update the Slack integration to use `PresetUrlOption.MAPPING` soon
